engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+		Name:             obj.Name(),        // string
+		ServerType:       serverType,        // *ServerType
+		Image:            image,             // *Image
+		SSHKeys:          keylist,           // []*SSHKey
+		Location:         nil,               // *Location
+		Datacenter:       datacenter,        // *Datacenter
+		UserData:         "",                // string
+		StartAfterCreate: &startAfterCreate, // *bool
+		Labels:           nil,               // map[string]string
+		Automount:        &automount,        // *bool
+		Volumes:          nil,               // []*Volume
+		Networks:         nil,               // []*Network
+		Firewalls:        nil,               // []*ServerCreateFirewall
+		PlacementGroup:   nil,               // *PlacementGroup
+	}
+	// hcloud-go provides basic validation, but this can still miss problems!
+	// TODO: add tests? If issues come up, add checks to Validate.
+	if err := hcloud.ServerCreateOpts.Validate(obj.serverconfig); err != nil {
+		return errwrap.Wrapf(err, "invalid serverconfig")
+	}
+	return nil
+}
+
+// getRescueKeys builds a list of keys to be enabled for rescue mode over SSH.
+// ServerRescueSSHKeys provides the selected keys as []string by name. The
+// corresponding data is collected with the Hetzner client (if valid). The
+// resulting keylist is formatted as []*hcloud.SSHKey for use with hcloud, and
+// saved for later use in private field rescueKeys.
+// TODO: standardize this so that it can also be used for serverconfig keys.
+func (obj *HetznerVMRes) getRescueKeys(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("getRescueKeys()")
+	}
+	var keylist []*hcloud.SSHKey
+	for _, keyname := range obj.ServerRescueSSHKeys {
+		key, _, err := obj.client.SSHKey.GetByName(ctx, keyname)
+		if err != nil {
+			return errwrap.Wrapf(err, "SSHKey GetByName(%s) failed", keyname)
+		}
+		if key == nil {
+			return fmt.Errorf("unknown keyname: %s", keyname)
+		}
+		if obj.init.Debug {
+			obj.init.Logf("append known keyname: %s", key)
+		}
+		keylist = append(keylist, key)
+	}
+	obj.rescueKeys = keylist
+	return nil
+}
+
+// waitFor provides a general function that returns nil when the provided exit
+// function is satisfied. It retries every WaitInterval until the condition is
+// satisfied. It can exit early in case the WaitTimeout is reached, the context
+// is cancelled or an error occurs. The exit function must be of type
+// HetznerVMCondition, and check a well-defined condition for the resource.
+func (obj *HetznerVMRes) waitFor(ctx context.Context, exitFunc HetznerVMCondition) error {
+	if obj.init.Debug {
+		obj.init.Logf("waitFor()")
+	}
+	var err error = nil
+	conditionSatisfied := false
+	timeout := time.After(time.Duration(obj.WaitTimeout) * time.Second)
+	for {
+		// Get up-to-date serverinfo.
+		if err := obj.getServerUpdate(ctx); err != nil {
+			return errwrap.Wrapf(err, "failed serverupdate request")
+		}
+		// Check if the provided exit condition is satisfied.
```

Comment:
You should add a comment here saying:

"The exitFunc does not take a content because it is guaranteed to never block or run for very long."

If this is not true, then of course it should take a context! And functions that can't use the context because they're very short and don't block should just ignore the using of that context of course!

Get it?

Reply:
Agreed and understood. Still need to adjust those little checkCondition functions, but will keep it in mind.